File: leetcode/Stack/leetcode-0085/85.go
package stack

func maximalRectangle(matrix [][]byte) int {
	if matrix == nil || len(matrix) == 0 || len(matrix[0]) == 0 {
		return 0
	}
	res := 0
	row, col := len(matrix), len(matrix[0])
	height := make([]int, col+1)
	for i := 0; i < row; i++ {
		for j := 0; j < col; j++ {
			if matrix[i][j] == '1' {
				height[j]++
			} else {
				height[j] = 0
			}
		}
		local := maxRectangle(height)
		if local > res {
			res = local
		}
	}
	return res
}

func maxRectangle(height []int) int {
	if height == nil || len(height) == 0 {
		return 0
	}

	i, res := 0, 0
	s := []int{}
	for i < len(height) {
		if len(s) == 0 || height[s[len(s)-1]] <= height[i] {
			s = append(s, i)
			i++
		} else {
			tmp := s[len(s)-1]
			s = s[:len(s)-1]
			local := 0
			if len(s) == 0 {
				local = i * height[tmp]
			} else {
				local = height[tmp] * (i - s[len(s)-1] - 1)
			}
			if local > res {
				res = local
			}
		}
	}
	return res
}

File: leetcode/Stack/leetcode-0085/85.java
class Solution {
    public int maximalRectangle(char[][] matrix) {
        if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
            return 0;
        }
        int row = matrix.length;
        int col = matrix[0].length;
        int[] height = new int[col + 1];
        int res = 0;
        for (int i = 0; i < row; i++) {
            for (int j = 0; j < col; j++) {
                if (matrix[i][j] == '1') {
                    height[j]++;
                } else {
                    height[j] = 0;
                }
            }
            res = Math.max(res, maxRectangle(height));
        }
        return res;
    }

    private int maxRectangle(int[] height) {
        if (height == null || height.length == 0) {
            return 0;
        }

        int i = 0, res = 0;
        Stack<Integer> s = new Stack<>();
        while (i < height.length) {
            if (s.isEmpty() || height[s.peek()] <= height[i]) {
                s.push(i++);
            } else {
                int tmp = s.pop();
                res = Math.max(res, height[tmp] * (s.isEmpty() ? i : i - s.peek() - 1));
            }
        }
        return res;
    }
}